lib/cloud-operation/communication/communication-channel-base.ts

```diff
@@ -0,0 +1,120 @@
+import { EventEmitter } from "events";
+import { CloudOperationMessageTypes, CloudCommunicationChannelExitCodes, CloudCommunicationEvents } from "../../constants";
+
+export abstract class CommunicationChannelBase extends EventEmitter implements ICloudCommunicationChannel {
+	private static readonly PING_INTERVAL: number = 10000;
+	private static readonly MISSING_PING_REPLIES_COUNT: number = 6;
+
+	protected isConnected: boolean;
+	private pingInterval: NodeJS.Timer;
+	private echoSequenceNumber: number = 0;
+	private echoReplySequenceNumber: number = 0;
+
+	private handshakeCompleteResolve: (value?: PromiseLike<void>) => void;
+
+	constructor(protected cloudOperationId: string,
+		protected data: ICloudCommunicationChannelData<any>,
+		protected $logger: ILogger) {
+		super();
+	}
+
+	public async connect(): Promise<void> {
+		if (!this.isConnected) {
+			await this.connectCore();
+			this.isConnected = true;
+			await this.handshake();
+			this.startPing();
+		}
+	}
+
+	public async sendMessage<T>(message: ICloudOperationMessage<T>): Promise<string> {
+		if (this.isConnected) {
+			const result = await this.sendMessageCore(message);
+
+			return result;
+		}
+
+		throw new Error("Communication channel not connected");
+	}
+
+	public async close(code: number, reason?: string): Promise<void> {
+		if (!this.isConnected) {
+			return;
+		}
+
+		try {
+			if (this.pingInterval) {
+				clearInterval(this.pingInterval);
+				this.pingInterval = null;
+			}
+
+			const stopMsg: ICloudOperationMessage<ICloudOperationStop> = {
+				type: CloudOperationMessageTypes.CLOUD_OPERATION_STOP,
+				cloudOperationId: this.cloudOperationId,
+				body: {
+					code,
+					reason
+				}
+			};
+			await this.sendMessage<ICloudOperationStop>(stopMsg);
+			await this.closeCore(code, reason);
+		} catch (err) {
+			this.$logger.trace(err);
+		}
+
+		this.isConnected = false;
+		this.emit(CloudCommunicationEvents.CLOSE, code, reason);
+	}
+
+	protected abstract connectCore(): Promise<void>;
+	protected abstract closeCore(code: number, reason?: string): Promise<void>;
+	protected abstract sendMessageCore<T>(message: ICloudOperationMessage<T>): Promise<string>;
+
+	protected handleMessage(m: any): ICloudOperationMessage<any> {
+		let msg: ICloudOperationMessage<any>;
+		if (_.isBuffer(m)) {
+			msg = JSON.parse(m.toString("utf8"));
+		} else if (_.isString(m)) {
+			msg = JSON.parse(m);
+		} else {
+			msg = m;
+		}
+
+		if (msg.type === CloudOperationMessageTypes.CLOUD_OPERATION_SERVER_HELLO) {
+			this.handshakeCompleteResolve();
+		} else if (msg.type === CloudOperationMessageTypes.CLOUD_OPERATION_ECHO_REPLY) {
+			const body: ICloudOperationEcho = msg.body;
+			if (body.identifier === this.cloudOperationId) {
+				this.echoReplySequenceNumber = body.sequenceNumber;
```

Comment:
Does it matter which exactly is the sequenceNumber? That is, can we just reset the counter here?